core/src/builtin_widgets/key.rs

```diff
@@ -1,278 +1,148 @@
-/// fn_widget!{
-///   @ {
-///     pipe!($trigger;).map(move |_| fn_widget!{
-///       @Margin {
-///       margin: EdgeInsets::all(10.),
-///         @KeyWidget {
-///           key: "key",
-///           value: (),
-///           @Void {}
+/// }
+/// let w = fn_widget! {
+///   let id_gen = Stateful::new(0);
+///   let items = Stateful::new(vec![]);
+///   @Column {
+///     @FilledButton{
+///       on_tap: move |_| {
+///         $items.write().push(Item { id: *$id_gen });
+///         *$id_gen.write() += 1;
+///       },
+///       @ { "add item"}
+///     }
+///     @ { pipe!($items;).map(move |_| {
+///         move || {
+///           $items.clone().into_iter().map(move |item| {
+///             @KeyWidget {
+///               key: item.id as isize,
+///               @ { item.into_lazy_widget() }
+///             }
+///           })
 ///         }
-///       }
-///   })
+///       })
+///     }
 ///   }
 /// };
-// `KeyWidget` should not support `FatObj`, as this may cause the `KeyWidget` to be invisible to its
-// parent. `@KeyWidget { margin: EdgeInsets::all(10.) }` actually expands as `@Margin { @KeyWidget {
-// .. } }`.
-#[simple_declare]
-
-pub struct KeyWidget<V: 'static> {
-  pub key: Key,
-  #[declare(strict)]
-  pub value: V,
-  #[declare(skip)]
-  before_value: Option<V>,
-  #[declare(skip)]
-  has_successor: bool,
-}
-
-/// A trait for `keyWidget` that use to record information of the previous and
-/// next key widget.
-pub(crate) trait AnyKey: Any {
-  fn key(&self) -> Key;
-  /// Record the previous KeyWidget associated with the same key.
-  fn record_prev_key_widget(&self, key: &dyn AnyKey);
-  /// Record the next KeyWidget associated with the same key.
-  fn record_next_key_widget(&self, key: &dyn AnyKey);
-  fn as_any(&self) -> &dyn Any;
-}
-
-impl<T, V> AnyKey for T
-where
-  T: StateWriter<Value = KeyWidget<V>>,
-  V: Clone + PartialEq + 'static,
-{
-  fn key(&self) -> Key { self.read().key.clone() }
-
-  fn record_prev_key_widget(&self, key: &dyn AnyKey) {
-    assert_eq!(self.key(), key.key());
-    let Some(key) = key.as_any().downcast_ref::<Self>() else {
-      log::warn!("Different value type for same key.");
-      return;
-    };
-    self
-      .write()
-      .record_before_value(key.read().value.clone());
-  }
-
-  fn record_next_key_widget(&self, _: &dyn AnyKey) { self.silent().has_successor = true; }
-
-  fn as_any(&self) -> &dyn Any { self }
-}
-
-impl<'c, V: 'static + Default + Clone + PartialEq> ComposeChild<'c> for KeyWidget<V> {
-  type Child = Widget<'c>;
-  fn compose_child(this: impl StateWriter<Value = Self>, child: Self::Child) -> Widget<'c> {
-    let data: Box<dyn AnyKey> = Box::new(this);
-    child.attach_data(Box::new(Queryable(data)))
+/// App::run(w);
+/// ```
+#[simple_declare(stateless)]
+pub struct KeyWidget {
+  pub(crate) key: Key,
+}
+
+impl KeyWidget {
+  pub fn with_child<'c, const M: usize, W>(self, child: W) -> (Key, W)
```

Comment:
There is no need to retain the `W` in this case; simply use `Widget<'c>` instead.

Reply:
not, the child may not be the Widget<'c>, like 
@KeyWidget {
   key: id as isize,
   @Void { }
  }